Feature/attachment enhance

chrisminett opened the pull request:
I've tried to break this down into logical commits, but the basic premise for the changes is this:
- There was no dedicated `Attachment` class, other than that used to allow a local file to be dynamically added to an existing `Mail` object (ie. no attachment class was used from the Factory)
- A charset was sometimes incorrectly defined on an attachment
- An attachment was sometimes not recognised as an attachment (we were only looking for _disposition_, but inline attachments won't necessarily have this)

So the overall changes are:
- Split local file handling into `AttachmentLocal`, leaving `Attachment` to be the consistent representation for an attachment, both for the Factory and dynamic usage.
- Only the `Mail` class gets a default charset, everything else needs it to be explicitly defined
  - The Factory only sets a charset for Attachments if it's defined for the content, rather than the part
- _Disposition_ is optional for attachments, but _name_ is required

The varying attachment attribute combinations (name, disposition, charset) are tested by the extended Factory tests. This was the basis for all the other changes.

chrisminett committed:
Remove default charset from AbstractPart, only required for Mail

chrisminett committed:
Extract content-type setter as a re-usable trait

chrisminett committed:
Merge branch 'master' into feature/attachment-enhance

chrisminett committed:
Separate methods for handling local attachments into separate class

This leaves the 'Attachment' class free for a pure representation of an attachment

chrisminett committed:
Make attachment name required and disposition optional

chrisminett committed:
Switch to using the Attachment class for any content part given a name

chrisminett committed:
Merge branch 'master' into feature/attachment-enhance

# Conflicts:
#	tests/FactoryTest.php

qubit05 commented on src/Content/AttachmentLocal.php:
```diff
@@ -0,0 +1,46 @@
+ *
+ * @package Phlib\Mail
+ */
+class AttachmentLocal extends Attachment
```
Would this be better called FileAttachment or AttachmentFile? AttachmentLocal is confusing how it relates to an email attachment.

What's the difference between Attachment and AttachmentLocal. Is one Inline? If so, would naming it more clearly be better?

chrisminett replied:
How about `AttachmentFile` or `AttachmentFromFile`, and making the _filename_ a constructor parameter, so immutable?

This class is designed to help a developer create a new part that matches the `Attachment` "interface" (so extends `Attachment`) from a file on their local filesystem. Essentially it's a shortcut to:

``` php
$part = new Attachment(mime_content_type($filename));
$part->setContent(file_get_contents($filename));
$part->setName(basename($filename));
```

qubit05 replied:
Sounds more like a static create method than a separate class to me. You've almost said it yourself in that comment, `AttachmentFromFile` and "create a new part".

chrisminett replied:
Yeah, so `Attachment::createFromFile($filename)` instead makes sense. Only downside is that the file content would have to be loaded on construct, whereas at the moment it's left until the call to `getContent()`. Do you think that's OK? If so I'll make the change.

qubit05 replied:
Other than the concern about large files and memory issues. I think this should be fine for now.

chrisminett replied:
OK, all waiting on #4 then.

qubit05 commented on src/SetTypeTrait.php:
```diff
@@ -0,0 +1,23 @@
+ *
+ * @package Phlib\Mail
+ */
+trait SetTypeTrait
```
Not sure about the SetTypeTrait idea. I see why you're doing it.

The 'type' property should be part of the trait.

chrisminett replied:
I've the behaviour to use immutable MIME types in #4 so this branch isn't responsible for changing the construct params of un-related classes. Then I can update the implementation here.

qubit05 commented on src/SetTypeTrait.php:
```diff
@@ -0,0 +1,23 @@
+ *
+ * @package Phlib\Mail
+ */
+trait SetTypeTrait
```
Looking at it a bit further, could you confirm my thinking:

It feels like there shouldn't be a setType method. The only type you want to specify it is on the Mime class. Creating a `new Mime('something/unusual')` class with other subtypes having types defined in the class? Then no `setType` method needed.

chrisminett committed:
Merge branch 'master' into feature/attachment-enhance

# Conflicts:
#	src/Content/Content.php
#	src/Factory.php
#	src/Mime/Mime.php

chrisminett committed:
Change Attachment to use immutable MIME type.

Remove SetTypeTrait

chrisminett committed:
Add new static method to create Attachment from local file, rather than a separate class

chrisminett committed:
Add Attachment name as required construct param

chrisminett commented:
OK, updated.

The Trait is gone. MIME type is now set in the construct (like #4) and I've put the attachment name there too, as that's a required attribute.

The _Local_ class is gone, replaced by a static _create_ method - much cleaner.

qubit05 commented on src/Content/Content.php:
```diff
@@ -2,6 +2,8 @@
 
 namespace Phlib\Mail\Content;
 
+use Phlib\Mail\SetTypeTrait;
```
Remove

qubit05 commented on src/Mime/Mime.php:
```diff
@@ -2,6 +2,8 @@
 
 namespace Phlib\Mail\Mime;
 
+use Phlib\Mail\SetTypeTrait;
```
Remove

qubit05 commented:
Why aren't the tests failing with the now unknown SetTypeTrait?

chrisminett committed:
Remove unused namespace import

chrisminett commented:
Seems as though it's because they were only namespace imports
